catkin_ws/src/global_perception/scripts/GlobalFuser.py

```diff
@@ -1,74 +1,167 @@
+                    
+                
+                if not found_match and min_gate_value > BBoxPredictor.new_track_threshold:
+                    # its a new id to add to the tracker
+                    id_key = str(bbox.value)
+                    print("new id to tracker id: " + id_key)
+                    self.tracked_bboxes[id_key] = {}
+                    self.tracked_bboxes[id_key]['timesteps_missed'] = 0
+                    self.tracked_bboxes[id_key]['latest_bbox'] = bbox
+                    bbox_pos = [bbox.pose.position.x,
+                                0,
+                                bbox.pose.position.y,
+                                0,
+                                bbox.pose.position.z,
+                                0]
+                    self.tracked_bboxes[id_key]['kf'] = BBoxPredictor(bbox_pos)
+                    ids_updated.append(id_key)
+
+        # check all the ids not updated, update with prediction instead
+        ids_not_updated = list(set(self.tracked_bboxes.keys()) - set(ids_updated))
+        for id in ids_not_updated:
+            self.tracked_bboxes[id]['timesteps_missed'] += 1
+            if self.tracked_bboxes[id]['timesteps_missed'] > TIME_THRESHOLD:
+                pass
+                # bbox has been gone too long delete it from tracker
+                print("deleting id " + str(id))
+                del self.tracked_bboxes[id]
+            else:
+                # else use its prediction for now
+                self.tracked_bboxes[id]['kf'].predict()
+                predicted_pos = self.tracked_bboxes[id]['kf'].get_posn() 
+                self.tracked_bboxes[id]['latest_bbox'].pose.position.x = predicted_pos[0]
+                self.tracked_bboxes[id]['latest_bbox'].pose.position.y = predicted_pos[1]
+                self.tracked_bboxes[id]['latest_bbox'].pose.position.z = predicted_pos[2]
+            
+        # at this point everything left in tracker should have its latest bbox published
+        fused_buffer = BoundingBoxArray()
+        fused_buffer.header.frame_id = 'map'
+        for id in self.tracked_bboxes.keys():
+            fused_buffer.boxes.append(self.tracked_bboxes[id]['latest_bbox'])
+        
+        self.pub.publish(fused_buffer)
 
 
-    def nodeCallback(self, data_n1, data_n2, data_n3):
-        # create new bbox to put in global buffer
-        self.master_buffer.node1_buffer.append(data_n1)
-        self.master_buffer.node2_buffer.append(data_n2)
-        self.master_buffer.node3_buffer.append(data_n3)
-        # self.fuserRun50ms(data_n1, data_n2, data_n3)
+    def nodeCallback(self, data_n1):
+        # Add all bboxes to raw buffer
+        self.master_buffer['node1'].header = data_n1.header
+        self.master_buffer['node1'].boxes = data_n1.boxes
+        # print(len(data_n1.boxes))
+        # if len(data_n1.boxes) == 3:
+        #     print("found it")
+        # self.master_buffer['node2'].header = data_n2.header
```

Comment:
Should node 2 be uncommented?

Reply:
No, this is because I got rid of time syncing

Reply:
This does need to be fixed tho because the buffer is just being overwritten and we could lose messages